src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -729,4 +729,11 @@ object messages {
            | - auxiliary constructors specify the implicit value
            |"""
   }
+
+  case class NoClassOrObjectAfterCase()(implicit ctx: Context)
+  extends Message(27) {
+    val kind = "Syntax"
+    val msg = hl"Only ${"class"} or ${"object"} are allowed after ${"case"}."
+    val explanation = ""
```

Comment:
Yes, `case trait` is just an example of what the user can write. What I'm trying to say is that the user should (if possible) be given the right solution to the problem.

I think that `msg` saying: `"Only ${"class"} or ${"object"} are allowed after ${"case"}."` is perfectly valid and sufficient.

And I think you should go for your second suggestion :) the point of having better error messages is indeed that they should be better! Do a lookahead at the next token and see if you can give some meaningful help in the `explanation` val. Sounds good?